src/window-menu.cpp

```diff
@@ -0,0 +1,147 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+
+#include <wayfire/core.hpp>
+#include <wayfire/view.hpp>
+#include <wayfire/plugin.hpp>
+#include <wayfire/output.hpp>
+#include <wayfire/signal-definitions.hpp>
+#include <wayfire/workspace-manager.hpp>
+#include <wayfire/output-layout.hpp>
+
+class wayfire_window_menu : public wf::plugin_interface_t
+{
+    /* The command should be set to a client that shows a menu window. */
+    wf::option_wrapper_t<bool> ignore_client_position{
+        "window-menu/ignore_client_position"};
+    wf::option_wrapper_t<std::string> command{"window-menu/command"};
+    wf::option_wrapper_t<std::string> app_id{"window-menu/app_id"};
+    wf::point_t position_offset;
+    wayfire_view menu_view;
+
+  public:
+    void init() override
+    {
+        printf("window-menu: init\n");
+        grab_interface->name = "window-menu";
+        grab_interface->capabilities = 0;
+
+        output->connect_signal("view-mapped", &view_mapped);
+        output->connect_signal("view-unmapped", &view_unmapped);
+        output->connect_signal("view-show-window-menu", &show_window_menu);
+        wf::get_core().connect_signal("pointer_button", &on_button);
+    }
+
+    wf::signal_connection_t view_mapped{[this] (wf::signal_data_t *data)
+        {
+            auto view = get_signaled_view(data);
+            if (view->get_app_id() == std::string(app_id))
+            {
+                view->set_decoration(nullptr);
+
+                /* Move to the respective output */
+                wf::get_core().move_view_to_output(view, output, false);
+
+                if (ignore_client_position)
+                {
+                    /* ignore client suggested position */
+                    auto position = output->get_cursor_position();
+                    view->move(position.x, position.y);
+                } else
+                {
+                    auto vg = view->get_wm_geometry();
+                    auto position = wf::point_t{vg.x, vg.y} + position_offset;
+                    view->move(position.x, position.y);
+                }
+
+                /* Place above other views */
+                output->workspace->add_view(view, wf::LAYER_UNMANAGED);
+
+                menu_view = view;
+                ((wf::view_mapped_signal*)data)->is_positioned = true;
+            }
+        }
+    };
+
+    wf::signal_connection_t on_button{[=] (wf::signal_data_t *data)
+        {
+            auto ev = static_cast<
+                wf::input_event_signal<wlr_event_pointer_button>*>(data);
+
+            if (ev->event->state != WLR_BUTTON_PRESSED)
+            {
+                return;
+            }
+
+            auto oc = output->get_cursor_position();
+            if (menu_view &&
+                !(menu_view->get_wm_geometry() & wf::point_t{(int)oc.x, (int)oc.y}))
```

Comment:
better to use for menu->map_input_coordinates() if you want to support a menu with submenus

Reply:
actually, if you had submenus, you'd need to look for popups of this main view (they'd be different popups entirely) ..

Maybe you should look at whether the cursor focus is the same wl_client as the menu?